### DISCUSSION TITLE
how read excel file to change json

### DISCUSSION THREAD

**Discussion by jaesungbong** [exceljs/exceljs#2479] [Q&A]
I have to change excel file to json

but there is no example source code.

so I found this source code.

but there is error 'Chunk must be one of type String, Buffer or StringBuf.'

What did I do wrong?

```javascript
<input type="file" className="form-control" id="file" onChange={this.parseExcel} />

parseExcel(e.target.files[0]).then((res) => {
      console.log(res)
})

parseExcel(fileContents) {
    return new Promise((resolve, reject) => {
        const reader = new FileReader();
        const wb = new ExcelJS.Workbook();
        reader.readAsArrayBuffer(fileContents);
            reader.onload = () => {
                const buffer = reader.result;
                wb.xlsx.load(buffer as Buffer).then(workbook => {
                    console.log(workbook, 'workbook instance')
            })
        }
    });
}
```

**Comment by Siemienik** [comment]
@jaesungbong 
I will recommend to you best solution what I know:). You may use `xlsx-import`, a library that I wrote some time ago and I'm using it currently a lot :)

the `xlsx-import` uses exceljs internally.

https://github.com/Siemienik/xlsx-import

`npm install xlsx-import --save`



```javascript
import ExcelJsImporterFactory from 'xlsx-import/lib/ImporterFactory';

const filePath = "./excel-file.xlsx";

const factory = new ImporterFactory(); 
const importer = await  factory.From(filePath);
return JSON.stringify(importer.GetAllItems({
            type: 'list',
            worksheet: 'Favourites',
            columns: [
                {
                    index: 1, // column index (1,2,3....n) 
                    key: 'Id', // output item's field
                    mapper: (v: string) => Number.parseInt(v) // nullable, for transformating values
                },
                {index: 2, key: 'Title'},
                {index: 5, key: 'Author'}, 
            ],
            rowOffset: 1, //offset header row
        }));
```

**Comment by Siemienik** [comment]
@guyonroche may I please you to add information about my import library to our readme?
What do you think about that?

**Comment by TomaszG** [comment]
> but there is error 'Chunk must be one of type String, Buffer or StringBuf.'

This was already raised there: https://github.com/exceljs/exceljs/issues/1098

> use xlsx-import, a library that I wrote some time ago and I'm using it currently a lot :)
the xlsx-import uses exceljs internally.

Yeah, it uses exceljs 3.5.0 (according to its package.json) which is free from the issue raised in #1098. I wouldn't call it "solution", @jaesungbong can achieve the same by downgrading exceljs to 3.5 (or 3.6.1 where it also worked fine).

**Comment by Siemienik** [comment]
`3.5.0` and `^3.5.0` make a difference. You're right, it's not only a solution, but it's a proposal way to much easier achieve this goal :)

> If you see ^1.0.2 it means to install version 1.0.2 or the latest minor or patch version such as 1.1.0.
https://michaelsoolee.com/npm-package-tilde-caret/

**Comment by lizhming** [comment]
@TomaszG 
I've tried to use 3.5 or 3.6.1, but I faced this error.
What's wrong with that?
```
app.js:8867 Uncaught (in promise) Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html
    at u.readEndOfCentral (app.js:8867)
    at u.load (app.js:8867)
    at app.js:8867
```

**Comment by Siemienik** [comment]
@lizhming did you read xlsx or xls file?

**Comment by lizhming** [comment]
I read xls file. this isn't supported xls file?

**Comment by Siemienik** [comment]
@lizhming unluckily not - xls is a binary file, 
exceljs handle only xlsx files (it's a zip file with XML inside).
